### PR TITLE
Feature majority ltf

### PR DESCRIPTION
**Pull request by MrM0nkey**
This pr implements a ltf array majority vote arbiter puf.

This pr implements a majority vote puf based on the LTFArray class.
I refactored NoisyLTFArray because it adds the same noise on each challenge https://github.com/nils-wisiol/pypuf/blob/feature_majority_ltf/pypuf/simulation/arbiter_based/ltfarray.py#L561-L562 . If this behavior is desired I will change my code in order to use the NoisyLTFArray.

### CODE REVIEW

File: pypuf/simulation/arbiter_based/ltfarray.py [hunk 1]
```diff
@@ -557,5 +558,44 @@ def ltf_eval(self, inputs):
         Random numbers are drawn from the PRNG instance generated when
         initializing the NoisyLTFArray.
         """
-        noise = self.random.normal(loc=0, scale=self.sigma_noise, size=(1, self.k))
-        return super().ltf_eval(inputs) + noise
+        evaled_inputs = super().ltf_eval(inputs)
+        noise = self.random.normal(loc=0, scale=self.sigma_noise, size=shape(evaled_inputs)[0:2])
```
Comment (nils-wisiol) [inline pypuf/simulation/arbiter_based/ltfarray.py]: `size=(len(inputs), self.k)` is easier to understand than `shape(evaled_inputs)[0:2]`, but for correct implementation of `ltf_eval` equivalent. (Also, why use `[0:2]` on a length 2 array?)

This change should be in a seperate bugfix commit.

File: pypuf/simulation/arbiter_based/ltfarray.py [hunk 2]
```diff
@@ -557,5 +558,44 @@ def ltf_eval(self, inputs):
...
+        return add(evaled_inputs, noise)
```
Comment (nils-wisiol) [inline pypuf/simulation/arbiter_based/ltfarray.py]: Please use infix notation unless you need additional parameters like `axis=`

File: test/test_simulation.py [hunk 3]
```diff
@@ -577,7 +577,66 @@ def test_ltf_eval(self):
                 random_instance=noise_prng_1,
             )
 
+            evaled_ltf_array = ltf_array.ltf_eval(transformed_inputs)
             assert_array_equal(
-                around(ltf_array.ltf_eval(transformed_inputs) + noise_prng_2.normal(size=(1, k)), decimals=10),
+                around(evaled_ltf_array + noise_prng_2.normal(loc=0, scale=1,
+                                                              size=shape(evaled_ltf_array)[0:2]), decimals=10),
```
Comment (nils-wisiol) [inline test/test_simulation.py]: please update to `size=(N,k)` (cf my review comment to the corresponding implementation)

**Comment by MrM0nkey** [comment]
### PR Checklist

- [ ] use ```size=(len(inputs), self.k)``` instead of shape
- [ ] bugfix commit for noise bug
- [ ] use ```+``` instead of ```add```
- [ ] rename ```MajorityLTFArray``` to ```SimulationMajorityLTFArray```
- [ ] remobe ```sigma_noise``` unused member variable
- [ ] use ```sum``` instead of ```add.reduce```
- [ ] increase the number of CRPs to 16
- [ ] use utils to generate inputs
- [ ] fix wording "This" to "These"

File: test/test_simulation.py [hunk 4]
```diff
@@ -577,7 +577,66 @@ def test_ltf_eval(self):
...
+
+        transformed_inputs = LTFArray.transform_id(
+            RandomState(seed=0xBAADA555).choice([-1, +1], (crp_count, n)),
+            k
+        )
+
+        ltf_array_result = around(ltf_array.eval(transformed_inputs), decimals=10)
+        mv_noisy_ltf_array_result = around(mv_noisy_ltf_array.eval(transformed_inputs), decimals=10)
+        # This test checks if the output is different cause by noise
+        self.assertFalse(array_equal(ltf_array_result, mv_noisy_ltf_array_result), 'This arrays must be different')
```
[the remark this answers is not in the document]
Comment (MrM0nkey) [reply test/test_simulation.py]: This is just one test why should I write "These"?

[the remark this answers is not in the document]
Comment (nils-wisiol) [reply test/test_simulation.py]: sorry I don't know what I was thinking xD

File: pypuf/simulation/arbiter_based/ltfarray.py [hunk 5]
```diff
@@ -557,5 +558,43 @@ def ltf_eval(self, inputs):
...
-        noise = self.random.normal(loc=0, scale=self.sigma_noise, size=(1, self.k))
-        return super().ltf_eval(inputs) + noise
+        evaled_inputs = super().ltf_eval(inputs)
+        noise = self.random.normal(loc=0, scale=self.sigma_noise, size=(len(evaled_inputs), self.k))
+        return evaled_inputs + noise
+
+
+class SimulationMajorityLTFArray(NoisyLTFArray):
+    """
+    This class provides a majority vote version of the LTFArray.
```
Comment (benj-zen) [inline pypuf/simulation/arbiter_based/ltfarray.py]: version of the NoisyLTFArray

File: test/test_simulation.py [hunk 6]
```diff
@@ -577,7 +577,96 @@ def test_ltf_eval(self):
...
+            weight_array=weigth_array,
+            transform=LTFArray.transform_id,
+            combiner=LTFArray.combiner_xor
+        )
+
+        inputs = array(list(tools.random_inputs(n, crp_count , random_instance=RandomState(seed=0xDEADDA7A))))
+
+        ltf_array_result = ltf_array.eval(inputs)
+        mv_noisy_ltf_array_result = mv_noisy_ltf_array.eval(inputs)
+        # This test checks if the output is different cause by noise
```
Comment (benj-zen) [inline test/test_simulation.py]: because of the noise

File: test/test_simulation.py [hunk 7]
```diff
@@ -577,7 +577,96 @@ def test_ltf_eval(self):
...
+        self.assertFalse(array_equal(ltf_array_result, mv_noisy_ltf_array_result), 'This arrays must be different')
```
Comment (benj-zen) [inline test/test_simulation.py]: These arrays

File: test/test_simulation.py [hunk 8]
```diff
@@ -577,7 +577,96 @@ def test_ltf_eval(self):
...
+            weight_array=weigth_array,
+            transform=LTFArray.transform_id,
+            combiner=LTFArray.combiner_xor,
+            sigma_noise=1,
+            random_instance_noise=noise_prng,
+            vote_count=vote_count,
+        )
+        # This checks if the majority vote works
+        mv_noisy_ltf_array_result = mv_noisy_ltf_array.eval(inputs)
+        assert_array_equal(mv_noisy_ltf_array_result,ltf_array_result)
```
Comment (benj-zen) [inline test/test_simulation.py]: whitespace

File: test/test_simulation.py [hunk 9]
```diff
@@ -577,7 +577,96 @@ def test_ltf_eval(self):
...
+        """
+        noise_prng = RandomState(seed=0xC0FFEE)
+        weight_prng1 = RandomState(seed=0xBADA55)
+        crp_count = 16  # number of random inputs per test set
+        n = 8
+        k = 2
+        mu = 0
+        sigma = 1
+        vote_count = 1
+        weigth_array = LTFArray.normal_weights(n, k, mu, sigma, weight_prng1)
```
Comment (benj-zen) [inline test/test_simulation.py]: Typo: weight_array

**Comment by nils-wisiol** [comment]
README needs update to reflect changes from acb4cb71ee1fe3a310e45f15c4bafdc90e3003c2, otherwise this lgtm

**Pull request merged**